### ISSUE TITLE
stop an Application but keep the application definition

### ISSUE THREAD

**Issue by rjdp** [argoproj/argo-cd#6079]
# Summary

Currently if i need to remove and application the application definition also gets deleted

# Motivation

Adding an application again has some overhead/toil.
I would like I could stop and application which undeploys the deployment but argo application definition should stay so i can reactivate

# Proposal

deactivate and application undeploys application and reactivating deploys it again

**Comment by RexHuang** [comment]
yes, it is useful, pls improve.

**Comment by BartoszGiza** [comment]
Yes I agree. I was looking for such option but without luck :(

**Comment by crenshaw-dev** [comment]
Can you elaborate on the use case? I think my first attempt at doing this would be to 1) disable autosync, 2) manually delete all resources managed by the app and 3) use sync windows to disable syncing.

But that's admittedly a bit hackier than a simple "off" button.

**Comment by geisterfurz007** [comment]
One possible usecase would be configuring an initial setup for example. I was setting up Argo for a project recently and had to fiddle with quite a few settings before the deployment finally worked out. The page wasn't supposed to be live though, so I wanted to stop the deployment again, just comitting some code changes necessary for everything to work while not losing the application's configuration in argo. I had to redo everything again recently which has the potential of misconfiguring again.

The workaround of disabling autosync and deleting resources would probably work but I guess requesting this as "native" feature for argo is not entirely invalid just as you said!

**Comment by QuinnBast** [comment]
I agree, it would be very nice to see this added. Sometimes we just want to completely remove applications from our cluster and there is no easy way to do that. Clicking the "Delete" button in the GUI deletes the entire ArgoCD application, which means that the Application completely disappears from the GUI. In order to get it back, we need to `kubectl apply -f ourArgoApplications.yaml` which is a pain. We've started to just delete apps by deleting the namespace they were deployed into, but this doesn't work for everything, especially apps that deploy to the same namespace, or use the default namespace.

There should be an option in Argocd to just remove the generated manifests from the cluster without removing the Application from Argo. I would initially thought that is what the "Delete" button in the GUI would do, but it does more than that.

**Comment by geisterfurz007** [comment]
I have some spare time over the holidays and would like to give this a shot. If some maintainer could give me a hint on where to start and possibly how to test out proper functionality of the feature, that would be appreciated!

**Comment by 12ain** [comment]
I searched for a lot of issues to implement similar functionality, but I found here that it doesn't seem to be implemented, and I think it would be a very useful feature

**Comment by geisterfurz007** [comment]
@crenshaw-dev since you upvoted my usecase comment about a year back, is this something Argo might be accepting a PR for? If so, I might give it a shot if you can give me a hint on where to look / what rough approach to follow.

**Comment by crenshaw-dev** [comment]
@geisterfurz007 I think I like the idea. There's another popular issue where the consensus seems to be that we need a feature to "scale down everything in the app that's scalable." Would that satisfy your use case? If so, I think it would be a great feature. https://github.com/argoproj/argo-cd/issues/3039#issuecomment-1593561051

Maybe something like this:

```yaml
apiVersion: argoproj.io/v1alpha1
kind: Application
spec:
  scaleDown: {}
```

Using a nilable object will allow us to add config later if necessary (e.g. a `matchLabels` block).

I have two concerns about this feature:
1) Argo CD needs a hard-coded list of "scalable" resources and how to scale them
2) This moves a very powerful feature out of git and into the Application - a user with `applications, update` access already has a lot of power, but this is a bit more

Maybe both of those things are fine? @jannfis curious what you think.

**Comment by geisterfurz007** [comment]
@crenshaw-dev I am not sure the two are necessarily equal. Take for instance a Service resource, which isn't scalable(I think). From what I understand, with your suggested approach, this Service would still exist, even if the whole application was scaled down which might not always be desirable. Yes, it would cover my usecase, but I can imagine plenty that wouldn't be covered. I would love to have a more user-friendly way of expressing the workaround you had mentioned in a previous comment which is essentially removing all associated resources with disabled auto-sync (ideally something that would override auto-sync so it wouldn't have to be considered each time an application was enabled or disabled).

**Comment by crenshaw-dev** [comment]
Gotcha. So the proposed feature is basically "delete everything and pause auto-sync." Would it also prevent manual syncs?

**Comment by QuinnBast** [comment]
> Would it also prevent manual syncs?

No, I imagine not, a manual sync would re-apply the manifests to the cluster. I imagine this just being a "Delete" button which goes and does a `kubectl delete -f ....` on the TARGET cluster which would be the inverse of a "Sync"; essentially just taking all the manifests down.

I think right now the main thing we need is some way to differentiate: "Delete Argo Application" vs. "Delete Live Manifests"

**Comment by crenshaw-dev** [comment]
We just expanded custom actions to support creating resources. How would you feel about a custom action on the Application resource which deletes the managed resources and then disables auto-sync? Another action could be designed to reverse those changes.

**Comment by dhirschfeld** [comment]
> We just expanded custom actions to support creating resources. How would you feel about a custom action on the Application resource which deletes the managed resources and then disables auto-sync? Another action could be designed to reverse those changes.

I think that sounds great. Would you need another action to reverse the delete? Couldn't you just do a manual sync to force a re-deploy of all the resources and toggle auto-sync back on?

**Comment by crenshaw-dev** [comment]
> Couldn't you just do a manual sync to force a re-deploy of all the resources and toggle auto-sync back on?

Yep! A "reverse" action would just be a convenience to bundle those two changes.

**Comment by crenshaw-dev** [comment]
Ah - neither `argocd app actions list` nor the UI show Application actions on the Application itself. The action only appears on the child resource in an app of apps.

That's probably easily fixed.

**Comment by fredleger** [comment]
Adding my thoughts on it since we do have one use case for it
After setup a new cluster under argocd control, we want to deploy some debug apps (something like http echo-header) to run some automated or manual tests.

Once the cluster has been validated, these apps are not useful anymore and can expose some security risks o we want to delete it. But from time to time they still can be useful to debug things (or prove some to other team), so we would like to re-enable it. Obviously that can go with a commit/revert cycle but this is not very handy to look in git history for the right commit.

I think that most of the use cases are also linked to the fact that in app-of-apps pattern, the auto-sync is inherited from parent so you can never break the auto-sync setting in tree leaves (maybe it's but i not in my experience so far). If auto-sync can be overridden, that may respond as well by desactivating it at the app level.

Happy to discuss it if there is need for clarification

**Comment by crenshaw-dev** [comment]
> in app-of-apps pattern, the auto-sync is inherited from parent

This isn't the case. Parent and child app sync settings are independent.

**Comment by fredleger** [comment]
> > in app-of-apps pattern, the auto-sync is inherited from parent
> 
> This isn't the case. Parent and child app sync settings are independent.

Need to investigate more why it has been an issue then
Still that don't change the thought that it can be useful to have a "stop / disable" button.
Thanks for the clarification

Fred

**Comment by crenshaw-dev** [comment]
btw, I believe App actions are now working, so the only additional feature needed would be `delete` support in custom actions. Then someone could write an action to delete all resources in an app.
